containers/src/Data/Map/Internal.hs

```diff
@@ -1330,18 +1533,29 @@ insertAlong q kx x (Bin sz ky y l r) =
 -- so instead we convert the value to a magical zero-width
 -- proxy that's ultimately erased.
 deleteAlong :: any -> BitQueue -> Map k a -> Map k a
-deleteAlong old !q0 !m = go (bogus old) q0 m where
+deleteAlong old !q0 !m = deleteAlong' (bogus old) q0 m
+
+deleteAlongNE :: any -> BitQueue -> NonEmptyMap k a -> Maybe (NonEmptyMap k a)
+deleteAlongNE old !q0 !m = nonEmpty $ deleteAlongFromNE' (bogus old) q0 m
+
 #ifdef USE_MAGIC_PROXY
-  go :: Proxy# () -> BitQueue -> Map k a -> Map k a
+deleteAlong' :: Proxy# () -> BitQueue -> Map k a -> Map k a
 #else
-  go :: any -> BitQueue -> Map k a -> Map k a
+deleteAlong' :: any -> BitQueue -> Map k a -> Map k a
 #endif
-  go !_ !_ Tip = Tip
-  go foom q (Bin _ ky y l r) =
-      case unconsQ q of
-        Just (False, tl) -> balanceR ky y (go foom tl l) r
-        Just (True, tl) -> balanceL ky y l (go foom tl r)
-        Nothing -> glue l r
+deleteAlong' !_ !_ Tip = Tip
+deleteAlong' foom q (NE t) = deleteAlongFromNE' foom q t
+
+#ifdef USE_MAGIC_PROXY
+deleteAlongFromNE' :: Proxy# () -> BitQueue -> NonEmptyMap k a -> Map k a
+#else
+deleteAlongFromNE' :: any -> BitQueue -> Map k a -> Map k a
```

Comment:
should the first `Map` be `NonEmptyMap`, check with GHC-7.6.3

Reply:
Yeah good catch
```suggestion
-deleteAlongFromNE' :: any -> BitQueue -> Map k a -> Map k a
+deleteAlongFromNE' :: any -> BitQueue -> NonEmptyMap k a -> Map k a
```